src/HealthChecks/HealthChecks/src/HealthCheckPublisherHostedService.cs

```diff
@@ -78,16 +89,53 @@ public Task StopAsync(CancellationToken cancellationToken = default)
             return Task.CompletedTask;
         }
 
-        _timer?.Dispose();
-        _timer = null;
+        if (_timers != null)
+        {
+            foreach (var timer in _timers)
+            {
+                timer.Dispose();
+            }
+
+            _timers = null;
+        }
 
         return Task.CompletedTask;
     }
 
-    // Yes, async void. We need to be async. We need to be void. We handle the exceptions in RunAsync
-    private async void Timer_Tick(object? state)
+    private List<Timer> CreateTimers()
     {
-        await RunAsync().ConfigureAwait(false);
+        var hcRegistrationsGrouped = new Dictionary<(TimeSpan Delay, TimeSpan Period), List<HealthCheckRegistration>>();
+        foreach (var hc in _healthCheckServiceOptions.Value.Registrations)
+        {
+            var timerOptions = GetTimerOptions(hc);
+            if (!hcRegistrationsGrouped.ContainsKey(timerOptions))
+            {
+                hcRegistrationsGrouped[timerOptions] = new List<HealthCheckRegistration>();
+            }
+            hcRegistrationsGrouped[timerOptions].Add(hc);
```

Comment:
I don't see where the `List<HealthCheckRegistration>` is being used from the dictionary. Can the dictionary just be a hashset instead?

Reply:
Nice catch, turned into hashset